addons/purchase/purchase.py

```diff
@@ -1644,10 +1644,10 @@ class product_product(osv.Model):
     _inherit = 'product.product'
     
     def _purchase_count(self, cr, uid, ids, field_name, arg, context=None):
-        Purchase = self.pool['purchase.order']
-        return {
-            product_id: Purchase.search_count(cr,uid, [('order_line.product_id', '=', product_id)], context=context) 
-            for product_id in ids
+        return { 
+            res['product_id'][0]: res['product_id_count']
+            for res in self.pool['purchase.order.line'].read_group(cr, uid, 
+                [('product_id', 'in', ids)], fields=['product_id', 'count(order_id)'], groupby=['product_id'])
```

Comment:
* I don't think count(order_id) is necessary, and I'm surprised it's even legal
* I'd suggest putting every parameter on its line
* maybe use new API calling style for read_group for easier forward port?